themes/default/content/docs/concepts/inputs-outputs/_index.md

```diff
@@ -1056,3 +1579,256 @@ public static Output<List<String>> split(Output<String> str) {
+		}),
+if err != nil {
+	return err
+}
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+var bucket = new Bucket("content-bucket", new BucketArgs
+{
+    Acl = "private",
+    Website = new BucketWebsiteArgs
+    {
+        IndexDocument = "index.html",
+        ErrorDocument = "404.html",
+    },
+});
+
+// TODO incorrect bucket policy build
+```
+
+{{% /choosable %}}
+{{% choosable language java %}}
+
+```java
+// TODO
+```
+
+{{% /choosable %}}
+
+{{< /chooser >}}
+
+Notice how the `apply` call is being used inside the JSON string. In this scenario, the `apply` call is happening during the build of the JSON string, which is too early in the Pulumi lifecycle. The value of the bucket ARN has not yet been resolved from the cloud provider API, so the JSON string cannot be built yet.
+
+The correct way of handling this scenario is to resolve the bucket ARN, _then_ build the JSON string. You can see an example of this in the (#outputs-and-json) section.
+
+Notice in that example how the JSON string is being built _inside_ the `apply` call to Pulumi. In logical order, this happens as:
+
+- Resolve the bucket ARN from the cloud provider
+- Then, build the JSON string with the "raw" value.
+
+### Resource Names
+
+When creating multiple resources, it can be tempting to use the output of one resource as the name to another:
+
+{{< chooser language "javascript,typescript,python,go,csharp,java" >}}
+
+{{% choosable language javascript %}}
+
+```javascript
+
+```
+
+{{% /choosable %}}
+{{% choosable language typescript %}}
+
+```typescript
+// NOTE: This example is not correct
+const bucket = new aws.s3.Bucket("my-bucket", {});
+
+const bucketPolicy = new aws.s3.BucketPolicy(bucket.name, {
+  // rest of bucket arguments go here
+});
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```python
+# NOTE: This example is not correct
+bucket = aws.s3.Bucket("my-bucket",)
+
+bucket_policy = aws.s3.BucketPolicy(
+    bucket.name,
+   # rest of bucket arguments go here
+)
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```go
+
+// NOTE: This example is not correct
+bucket, err := s3.NewBucket(ctx, "my-bucket", &s3.BucketArgs{})
+if err != nil {
+	return err
+}
+
+_, err = s3.NewBucketPolicy(ctx, bucket.Name, &s3.BucketPolicyArgs{
+	Bucket: bucket.ID(),
+	// rest of bucket arguments go here
+})
+if err != nil {
+	return err
+}
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```csharp
+var bucket = new Bucket("my-bucket", new BucketArgs{});
+
+// TODO incorrect bucket policy build
+```
+
+{{% /choosable %}}
+{{% choosable language java %}}
+
+```java
+// TODO
+```
+
+{{% /choosable %}}
+{{< /chooser >}}
+
+However, this pattern isn't possible because resource names need to be available and known at runtime in order to construct the URN of the resource.
+
+If you really wish to pass an output from one resource as the resource name of another resource, you will need to use a {{< pulumi-apply >}}.
```

Comment:
I think using `{{< pulumi-apply >}}` is not working outside of choosers, I get a broken link. Should we use `[apply](#apply)` in this document?